doi-requests-commons/src/main/java/no/unit/nva/doi/requests/service/impl/DynamoDBDoiRequestsService.java

```diff
@@ -0,0 +1,107 @@
+    public static final String INDEX_NAME = "INDEX_NAME";
+    public static final String DOI_REQUEST_STATUS_DATE = "doiRequestStatusDate";
+    public static final String PUBLISHER_ID = "publisherId";
+    public static final String ERROR_READING_FROM_TABLE = "Error reading from table";
+
+    private final Logger logger = LoggerFactory.getLogger(DynamoDBDoiRequestsService.class);
+    private final Table table;
+    private final Index index;
+    private final ObjectMapper objectMapper;
+
+    /**
+     * Constructor for DynamoDBDoiRequestsService.
+     *
+     * @param objectMapper  objectMapper
+     * @param table DynamoDB table
+     * @param index DynamoDB index
+     */
+    public DynamoDBDoiRequestsService(ObjectMapper objectMapper, Table table, Index index) {
+        this.objectMapper = objectMapper;
+        this.table = table;
+        this.index = index;
+    }
+
+    /**
+     * Constructor for DynamoDBDoiRequestsService.
+     *
+     * @param client    AmazonDynamoDB client
+     * @param objectMapper  objectMapper
+     * @param environment   environment reader
+     */
+    public DynamoDBDoiRequestsService(AmazonDynamoDB client, ObjectMapper objectMapper, Environment environment) {
+        String tableName = environment.readEnv(TABLE_NAME);
+        String indexName = environment.readEnv(INDEX_NAME);
+        DynamoDB dynamoDB = new DynamoDB(client);
+        this.table = dynamoDB.getTable(tableName);
+        this.index = table.getIndex(indexName);
+        this.objectMapper = objectMapper;
+    }
+
+    @Override
+    public List<DoiRequestSummary> findDoiRequestsByStatus(URI publisher, DoiRequestStatus status)
+        throws ApiGatewayException {
+        RangeKeyCondition rangeKeyCondition = new RangeKeyCondition(DOI_REQUEST_STATUS_DATE);
+        rangeKeyCondition.beginsWith(status.toString());
+        ItemCollection<QueryOutcome> outcome;
+        try {
+            outcome = index.query(PUBLISHER_ID, publisher.toString(), rangeKeyCondition);
+        } catch (Exception e) {
+            throw new DynamoDBException(ERROR_READING_FROM_TABLE, e);
+        }
+        return parseJsonToDoiRequestSummaries(outcome);
+    }
+
+    @Override
+    public List<DoiRequestSummary> findDoiRequestsByStatusAndOwner(URI publisher, DoiRequestStatus status, String owner)
+        throws ApiGatewayException {
+        return findDoiRequestsByStatus(publisher, status)
+            .stream()
+            .filter(doiRequestSummary -> doiRequestSummary.getPublicationOwner().equals(owner))
+            .collect(Collectors.toList());
+    }
+
+    private List<DoiRequestSummary> parseJsonToDoiRequestSummaries(ItemCollection<QueryOutcome> items) {
+        List<DoiRequestSummary> doiRequestSummaries = new ArrayList<>();
+        items.forEach(item -> toDoiRequestSummary(item).ifPresent(doiRequestSummaries::add));
+        return doiRequestSummaries;
+    }
+
+    protected Optional<DoiRequestSummary> toDoiRequestSummary(Item item) {
```

Comment:
It seems that we are using the database model as the interface model. Although this is not necessarily bad, it could mean that any change in the way we store data, could break our interface and that our clients would need to update the handling of their queries. Maybe the data transfer objects should be logically different from the database objects. It makes things a bit more complicated but it will allow us a ) to store more information in our database than we need to show to the user, b)  restructure the data in our database without the client noticing, c) have redundant fields for secondary indices if necessary, d) deal with possible inconsistencies of how Dynamo and Jackson serialize data types
Bonus: We can have the DTO models in a package that other services can import through gradle and if we change the interface, then the other services would only need to upgrade their dependencies.

Reply:
DoiRequestSummary is a pure Dto class, but maybe it is wrong to have it originate from the Service.